<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Расчет-количества-книг-вышедших-после-1-января-2000-года" data-toc-modified-id="Расчет-количества-книг-вышедших-после-1-января-2000-года-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Расчет количества книг вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Расчет-количества-обзоров-и-средней-оценки-для-каждой-книги" data-toc-modified-id="Расчет-количества-обзоров-и-средней-оценки-для-каждой-книги-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Расчет количества обзоров и средней оценки для каждой книги</a></span></li><li><span><a href="#Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определение-издательства,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Определение издательства, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определение-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определение-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определение автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Расчет-среднего-количества-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Расчет-среднего-количества-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></div>

**SQL**

**Описание проекта**
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В
свободное время жители городов больше не выходят на улицу, не посещают кафе
и торговые центры. Зато стало больше времени для книг. Это заметили
стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг
по подписке. Ваша первая задача как аналитика — проанализировать базу
данных.

**Цель исследования:** проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


## Загрузка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# выгрузим все таблицы через цикл
tabels = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for table in tabels:
    query = ''' SELECT * FROM {}; 
    '''.format(table)
    request = pd.io.sql.read_sql(query, con = engine)
    display('Таблица - {}'.format(table))
    display(request.head(10))
    display(request.info())

'Таблица - books'

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

'Таблица - authors'

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

'Таблица - publishers'

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

'Таблица - ratings'

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

'Таблица - reviews'

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

**Описание данных**

Имеем 5 таблиц:

Таблица `books` cодержит данные о книгах:
* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

Таблица `authors` cодержит данные об авторах:
* author_id — идентификатор автора;
* author — имя автора.

Таблица `publishers` cодержит данные об издательствах:
* publisher_id — идентификатор издательства;
* publisher — название издательства;

Таблица `ratings` cодержит данные о пользовательских оценках книг:
* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

Таблица `reviews` cодержит данные о пользовательских обзорах на книги:
* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя пользователя, написавшего обзор;
* text — текст обзора.

**Вывод**:

С данными все хорошо, пропусков нет, типы данных соответствуют описанию, названия столбцов в таблицах в нижнем регистре и соответствуют "змеиному стилю".

## Расчет количества книг вышедших после 1 января 2000 года

In [4]:
query_1 = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01';
'''

In [5]:
request_1 = pd.io.sql.read_sql(query_1, con = engine)
display('После 1 января 2000 года вышло {} книг'.format(request_1.iloc[0,0]))

'После 1 января 2000 года вышло 819 книг'

## Расчет количества обзоров и средней оценки для каждой книги

In [6]:
query_rev = '''
SELECT * FROM reviews;
'''
reviews = pd.io.sql.read_sql(query_rev, con = engine)

query_b = '''
SELECT * FROM books;
'''
books = pd.io.sql.read_sql(query_b, con = engine)

query_rat = '''
SELECT * FROM ratings;
'''
ratings = pd.io.sql.read_sql(query_rat, con = engine)

In [7]:
display('Количество уникальных book_id в таблице reviews {}'.format(len(reviews['book_id'].unique())))
display('Количество уникальных book_id в таблице books {}'.format(len(books['book_id'].unique())))
display('Количество уникальных book_id в таблице ratings {}'.format(len(ratings['book_id'].unique())))

'Количество уникальных book_id в таблице reviews 994'

'Количество уникальных book_id в таблице books 1000'

'Количество уникальных book_id в таблице ratings 1000'

У всех 3-х таблиц одинаковый столбец **book_id**, будем соединять по этому столбцу. Как видим в таблице **reviews** в столбце **book_id 994** значений, в остальных по **1 000** значений, поэтому мы сначала объединим **books** и **ratings**, а затем **LEFT JOIN** присоединим **reviews**.

In [8]:
query_2 = '''
SELECT b.title, AVG(rat.rating) AS avg_rating, COUNT(DISTINCT rev.review_id) AS cnt_review
FROM books AS b
JOIN ratings AS rat ON b.book_id = rat.book_id
LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
GROUP BY b.book_id
ORDER BY cnt_review DESC;
'''
request_2 = pd.io.sql.read_sql(query_2, con = engine)

In [9]:
request_2

title  avg_rating  cnt_review
0                              Twilight (Twilight  #1)    3.662500           7
1                                  Water for Elephants    3.977273           6
2                                     The Glass Castle    4.206897           6
3    Harry Potter and the Prisoner of Azkaban (Harr...    4.414634           6
4    The Curious Incident of the Dog in the Night-Time    4.081081           6
..                                                 ...         ...         ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel    3.666667           0
996                            The Natural Way to Draw    3.000000           0
997   The Cat in the Hat and Other Dr. Seuss Favorites    5.000000           0
998                          Essential Tales and Poems    4.000000           0
999  Disney's Beauty and the Beast (A Little Golden...    4.000000           0

[1000 rows x 3 columns]

**Вывод**

Топ-5 по количеству обзоров:
- Twilight (Twilight #1) **7**
- Water for Elephants	**6**
- The Glass Castle	**6**
- Harry Potter and the Prisoner of Azkaban 4.414634	**6**
- The Curious Incident of the Dog in the Night-Time **6**

У шести книг не было ревью вообще

## Определение издательства, которое выпустило наибольшее число книг толще 50 страниц

In [10]:
query_3 = '''
SELECT p.publisher, COUNT(book_id)
FROM publishers AS p
LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY COUNT(book_id) DESC
LIMIT 1;
'''
request_3 = pd.io.sql.read_sql(query_3, con = engine)

In [11]:
display('Издательство, выпустившее наибольшее число книг толще 50 листов, {}'.format(request_3.iloc[0,0]))

'Издательство, выпустившее наибольшее число книг толще 50 листов, Penguin Books'

## Определение автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [12]:
query_4 = '''
SELECT authors.author, AVG(rat.avg_rating) AS avg_rating
FROM books
LEFT JOIN (
        SELECT book_id, COUNT(rating_id) AS cnt_rating, AVG(rating) AS avg_rating
        FROM ratings
        GROUP BY book_id
        ) AS rat ON rat.book_id = books.book_id
LEFT JOIN authors ON authors.author_id = books.author_id
WHERE cnt_rating > 50
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
'''
request_4 = pd.io.sql.read_sql(query_4, con = engine)

In [13]:
display('Автор с самой высокой средней оценкой книг {}'.format(request_4.iloc[0,0]))

'Автор с самой высокой средней оценкой книг J.K. Rowling/Mary GrandPré'

## Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
query_5 = '''
SELECT ROUND(AVG(sub_text_cnt.text_cnt)) AS avg_text_cnt
FROM (SELECT COUNT(text) AS text_cnt
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING
                         COUNT(ratings) > 50)
      GROUP BY username
      ) AS sub_text_cnt  
'''
request_5 = pd.io.sql.read_sql(query_5, con = engine)

In [15]:
display('Среднее количество обзоров пользователей, которые поставили больше 50 оценок составляет {}'.format(request_5.iloc[0,0]))

'Среднее количество обзоров пользователей, которые поставили больше 50 оценок составляет 24.0'

**Вывод**

После 1 января 2000 года было выпущено **819** книг.

Топ-5 по количеству обзоров:
- Twilight (Twilight #1) **7**
- Water for Elephants	**6**
- The Glass Castle	**6**
- Harry Potter and the Prisoner of Azkaban 4.414634	**6**
- The Curious Incident of the Dog in the Night-Time **6**

Издательство, выпустившее наибольшее число книг толще 50 листов, **Penguin Books**

Автор с самой высокой средней оценкой книг **J.K. Rowling/Mary GrandPré**

Среднее количество обзоров пользователей, которые поставили больше 50 оценок составляет **24.0**